cpp/src/strings/convert/convert_urls.cu

```diff
@@ -222,30 +286,84 @@ std::unique_ptr<column> url_decode(
   size_type strings_count = strings.size();
   if (strings_count == 0) return make_empty_strings_column(stream, mr);
 
-  auto strings_column = column_device_view::create(strings.parent(), stream);
-  auto d_strings      = *strings_column;
+  auto d_offsets = strings.offsets().data<size_type>();
+  // use indices relative to the base column data so they can be compared to offset values
+  auto d_in_chars  = strings.chars().head<char>();
+  auto chars_bytes = strings.chars_size();
+  // determine index of first character in base column
+  size_type chars_start = 0;
+  if (strings.offset() != 0) {
+    chars_start = cudf::detail::get_value<size_type>(strings.offsets(), 0, stream);
+  }
+  size_type chars_end = chars_start + chars_bytes;
+
+  url_decode_escape_detector esc_detector{chars_end, d_in_chars};
+
+  // Count the number of URL escape sequences across all strings, ignoring string boundaries.
+  // This may count more sequences than actually are there since string boundaries are ignored.
+  size_type potential_esc_count =
+    thrust::count_if(rmm::exec_policy(stream),
+                     thrust::make_counting_iterator<size_type>(chars_start),
+                     thrust::make_counting_iterator<size_type>(chars_end),
+                     esc_detector);
+
+  if (potential_esc_count == 0) {
+    // nothing to replace, so just copy the input column
+    return std::make_unique<cudf::column>(strings.parent());
+  }
+
+  // create a vector of potential escape sequence positions
+  rmm::device_uvector<size_type> potential_esc_positions(potential_esc_count, stream);
+  auto d_potential_esc_positions = potential_esc_positions.data();
+  thrust::copy_if(rmm::exec_policy(stream),
+                  thrust::make_counting_iterator<size_t>(chars_start),
+                  thrust::make_counting_iterator<size_t>(chars_end),
+                  d_potential_esc_positions,
+                  esc_detector);
+
+  // Filter potential escape positions to actual positions by checking string boundaries.
+  // Skip counting the actual number to compute the optimal output buffer size since it is not
+  // expected to be significantly smaller in practice.
+  rmm::device_uvector<size_type> esc_positions(potential_esc_count, stream);
+  auto d_esc_positions = potential_esc_positions.data();
+  auto esc_pos_end     = thrust::copy_if(rmm::exec_policy(stream),
+                                     d_potential_esc_positions,
+                                     d_potential_esc_positions + potential_esc_count,
+                                     d_esc_positions,
+                                     url_decode_esc_position_filter{strings.size() + 1, d_offsets});
+  size_type esc_count  = esc_pos_end - d_esc_positions;
```

Comment:
This is possibly alot of working memory. I see a couple of options.
1. `potential_esc_positions` is only used to build `esc_positions` so you could possibly use something like [IILE](https://www.bfilipek.com/2016/11/iife-for-complex-initialization.html#iife) to initialize `esc_positions` instead and thereby free `potential_esc_positions` after `esc_positions` is created.
2. Rename `potential_esc_positions` back to `esc_positions` and use `thrust::remove_if` to remove the invalid positions in-place instead -- since the potential positions are always greater than or equal to the real positions. This will reduce the need for a 2nd buffer which more than likely would not be necessary anyway.

Reply:
Yeah, I had already refactored locally to free `potential_esc_positions` early, but it looks like `thrust::remove_if` for in-place update would be the way to go.  Thanks!